// redux-effects-fetch/lib/index.js
/**
 * Imports
 */

'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _isomorphicFetch = require('isomorphic-fetch');

var _isomorphicFetch2 = _interopRequireDefault(_isomorphicFetch);

/**
 * Action types
 */

var FETCH = 'EFFECT_FETCH';

/**
 * Fetch middleware
 */

function fetchMiddleware(_ref) {
  var dispatch = _ref.dispatch;
  var getState = _ref.getState;

  return function (next) {
    return function (action) {
      return action.type === FETCH ? (0, _isomorphicFetch2['default'])(action.payload.url, action.payload.params).then(checkStatus).then(deserialize, deserializeError) : next(action);
    };
  };
}

/**
 * Deserialize the request body
 */

function deserialize(res) {
  var header = res.headers.get('Content-Type');
  return header && header.indexOf('application/json') > -1 ? res.json() : res.text();
}

/**
 * Deserialize the request body, then return a
 * new rejected promise so the failure chain
 * stays failed.
 */

function deserializeError(res) {
  // new rejected Promise to go down failure chain
  return deserialize(res).then(function (body) {
    throw body;
  });
}

/**
 * Check the status and reject the promise if it's not in the 200 range
 */

function checkStatus(res) {
  if (res.status >= 200 && res.status < 300) {
    return res;
  } else {
    throw res;
  }
}

/**
 * Action creator
 */

function fetch() {
  var url = arguments.length <= 0 || arguments[0] === undefined ? '' : arguments[0];
  var params = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

  return {
    type: FETCH,
    payload: {
      url: url,
      params: params
    }
  };
}

/**
 * Exports
 */

exports['default'] = fetchMiddleware;
exports.fetch = fetch;
